fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -3,204 +3,546 @@
+            del function, chk
+
+        else:
+
+            def compute(
+                batch: Batch = batch,
+                chunk_id: int = chunk,
+                rank: int = dist_record.rank if dist_record is not None else -1,
+            ) -> Batch:
+                with record_function("chunk%d-rank%d" % (chunk_id, dist_record.rank)):
+                    result = self.module(*batch.tensors)
+                    if self.num_output is None:
+                        result = (result,)
+                return Batch(result, chunk_id)
+
+            task = Task(self.stream, compute=compute, finalize=None)
+            del compute
+
+        self.in_queue.put(task)
+
+        ok, payload = self.out_queue.get()
+
+        # Hold the first exception.
+        if exc_info is not None:
+            pass
+        elif not ok:
+            exc_info = cast(ExcInfo, payload)
+        else:
+            task, batch = cast(Tuple[Task, Batch], payload)
+
+            with use_device(self.device):
+                task.finalize(batch)
+
+            dist_record.batches[chunk] = batch
+
+        if exc_info is not None:
+            raise exc_info[0].with_traceback(exc_info[1], exc_info[2])
+
+    def forward_results(self, chunk: int, dist_record: DistributedPipelineRecord) -> None:
+        with use_stream(self.stream):
+            for user, input_idx, output_idx in dist_record.users:
+                v = dist_record.get_batch(chunk).value[output_idx]
+                dist_record.forwarded_phony[chunk][output_idx].append(user.remote().feed(chunk, input_idx, v))
+
+    def run_pipeline(self, dist_record_ref: rpc.RRef) -> Optional[Tensor]:
+        dist_record = dist_record_ref.local_value()
+        self.run(dist_record)
+
+        if not dist_record.users:
+            result = microbatch.gather(dist_record.batches)
+            assert len(result) == 1
+            result = result[0]
+            s0 = current_stream(result.device)
+            if is_cuda(s0):
+                # TODO. Investigate why this is needed and remove it if possible.
+                as_cuda(s0).synchronize()
+            return result
+
+        return None
+
+
+class MultiInputSequential(nn.Module):
+    """
+        A variation of nn.Sequential, that allows the first module in the sequence accepts
+        multiple inputs.
+    """
+
+    def __init__(self, *modules: nn.Module) -> None:
         super().__init__()
+        self.modules_list = modules
+
+    def forward(self, *inputs: Tuple[Tensor]) -> Tensor:  # type: ignore
+        input = self.modules_list[0](*inputs)
+        for module in self.modules_list[1:]:
+            input = module(input)
+        return input
+
+
+def RemoteSequential(rref_list: List[rpc.RRef]) -> MultiInputSequential:
+    return MultiInputSequential(*(r.local_value() for r in rref_list))
+
+
+def _split_module(g: PipelineModulesGraph) -> List[Tuple[List[int], rpc.RRef]]:
+    g.compute_output_users()
```

Comment:
This seems like too large a scope over which to use a single-letter variable name. Single-letter variable names should only be used in list comprehensions and short loops. Use `graph` instead?